### PR TITLE
rA-Merge: Rebellion Skill

### PR DESCRIPTION
**Pull request by MrKeiKun**
### Pull Request Prelude

- [x] I have followed [proper Hercules code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed
Merged rAthena's Rebellion Skills

**Note:**
Some skills fields are still not up-to-date, from upstream. Will them later.

**Issues addressed:** 

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

### CODE REVIEW

File: db/re/skill_db.conf [hunk 1]
```diff
@@ -31925,18 +31943,28 @@ skill_db: (
...
-			Lv9: 112
-			Lv10: 116
+			Lv1: 40
+			Lv2: 44
+			Lv3: 48
+			Lv4: 56
+			Lv5: 60
+		}
+		WeaponTypes: {
+			Rifle: true
```
Comment (MrKeiKun) [inline db/re/skill_db.conf:31960]: yeah mistake here.

File: src/map/battle.c [hunk 2]
```diff
@@ -2272,6 +2277,8 @@ static int battle_calc_skillratio(int attack_type, struct block_list *src, struc
 					break;
 				case GS_DESPERADO:
 					skillratio += 50 * (skill_lv-1);
+					if (sc &&  sc->data[SC_FALLEN_ANGEL])
```
Comment (EPuncker) [inline src/map/battle.c:2280]:
```suggestion
					if (sc && sc->data[SC_FALLEN_ANGEL])
```

File: db/re/skill_db.conf [hunk 3]
```diff
@@ -31984,31 +32026,47 @@ skill_db: (
...
 		Lv4: 13000
 		Lv5: 14000
-		Lv6: 15000
-		Lv7: 16000
-		Lv8: 17000
-		Lv9: 18000
-		Lv10: 19000
 	}
-	FixedCastTime: 0
+    Cooldown: 10000
```
Comment (EPuncker) [inline db/re/skill_db.conf:32051]:
```suggestion
	Cooldown: 10000
```

File: db/re/skill_db.conf [hunk 4]
```diff
@@ -31984,31 +32026,47 @@ skill_db: (
...
 	SkillInstances: 3
-	CoolDown: {
+	CastTime: {
+		Lv1: 2000
+		Lv2: 1600
+		Lv3: 1200
+		Lv4: 800
+		Lv5: 400
+	}
+    AfterCastActDelay: {
```
Comment (EPuncker) [inline db/re/skill_db.conf:32036]:
```suggestion
	AfterCastActDelay: {
```

File: db/re/skill_db.conf [hunk 5]
```diff
@@ -31956,24 +31984,38 @@ skill_db: (
...
+	AfterCastActDelay: 1000
+	CastTime: {
+		Lv1: 3000
+		Lv2: 2500
+		Lv3: 2000
+		Lv4: 1500
+		Lv5: 1000
+	}
+	FixedCastTime: 1000
+    Cooldown: 2000
```
Comment (EPuncker) [inline db/re/skill_db.conf:31996]:
```suggestion
	Cooldown: 2000
```

File: db/re/skill_db.conf [hunk 6]
```diff
@@ -31925,18 +31943,28 @@ skill_db: (
 	}
 	AttackType: "Weapon"
 	Element: "Ele_Weapon"
+	CastTime: 1000
+    AfterCastActDelay: 1000
```
Comment (EPuncker) [inline db/re/skill_db.conf:31947]:
```suggestion
	AfterCastActDelay: 1000
```

File: db/re/skill_db.conf [hunk 7]
```diff
@@ -31925,18 +31943,28 @@ skill_db: (
...
+    SkillData1: 120000
```
Comment (EPuncker) [inline db/re/skill_db.conf:31948]:
```suggestion
	SkillData1: 120000
```

File: db/re/skill_db.conf [hunk 8]
```diff
@@ -31925,18 +31943,28 @@ skill_db: (
...
+    Cooldown: 2000
```
Comment (EPuncker) [inline db/re/skill_db.conf:31949]:
```suggestion
	Cooldown: 2000
```

File: db/re/skill_db.conf [hunk 9]
```diff
@@ -31925,18 +31943,28 @@ skill_db: (
...
+    FixedCastTime: 2000
```
Comment (EPuncker) [inline db/re/skill_db.conf:31950]:
```suggestion
	FixedCastTime: 2000
```

File: db/re/skill_db.conf [hunk 10]
```diff
@@ -31984,31 +32026,47 @@ skill_db: (
...
+		Lv5: 400
+	}
+    AfterCastActDelay: {
+		Lv1: 2500
+		Lv2: 2000
+		Lv3: 1500
+		Lv4: 1000
+		Lv5: 500
+	}
+    SkillData1: 10000
```
Comment (EPuncker) [inline db/re/skill_db.conf:32043]:
```suggestion
	SkillData1: 10000
```

File: db/re/skill_db.conf [hunk 11]
```diff
@@ -31984,31 +32026,47 @@ skill_db: (
...
+    SkillData2: {
```
Comment (EPuncker) [inline db/re/skill_db.conf:32044]:
```suggestion
	SkillData2: {
```

File: db/re/skill_db.conf [hunk 12]
```diff
@@ -32066,25 +32126,43 @@ skill_db: (
...
+	}
+	CastTime: {
+		Lv1: 3000
+		Lv2: 2500
+		Lv3: 2000
+		Lv4: 1500
+		Lv5: 1000
+	}
+	Cooldown: 2000
+    FixedCastTime: 1000
```
Comment (EPuncker) [inline db/re/skill_db.conf:32144]:
```suggestion
	FixedCastTime: 1000
```

File: db/re/skill_db.conf [hunk 13]
```diff
@@ -32094,31 +32172,72 @@ skill_db: (
 	DamageType: {
 		NoDamage: true
 	}
+	CastTime: 1000
+    AfterCastActDelay: 1000
```
Comment (EPuncker) [inline db/re/skill_db.conf:32176]:
```suggestion
	AfterCastActDelay: 1000
```

File: db/re/skill_db.conf [hunk 14]
```diff
@@ -32094,31 +32172,72 @@ skill_db: (
...
+    SkillData1: {
```
Comment (EPuncker) [inline db/re/skill_db.conf:32177]:
```suggestion
	SkillData1: {
```

File: db/re/skill_db.conf [hunk 15]
```diff
@@ -32094,31 +32172,72 @@ skill_db: (
...
 	Requirements: {
 		SPCost: 45
+		WeaponTypes: {
+			Revolvers: true
+			Rifles: true
+			GatlingGuns: true
+			Shotguns: true
+			GrenadeLaunchers: true
+		}
+      SpiritSphereCost: 1
```
Comment (EPuncker) [inline db/re/skill_db.conf:32198]:
```suggestion
	SpiritSphereCost: 1
```

File: db/re/skill_db.conf [hunk 16]
```diff
@@ -32094,31 +32172,72 @@ skill_db: (
...
 	}
 	SkillInfo: {
-		NoCastSelf: true
+		IsCombo: true
 	}
 	AttackType: "Weapon"
 	Element: "Ele_Weapon"
+	SplashRange: 10
+	NumberOfHits: 1
+    SkillData1: 1500
```
Comment (EPuncker) [inline db/re/skill_db.conf:32218]:
```suggestion
	SkillData1: 1500
```

File: db/re/skill_db.conf [hunk 17]
```diff
@@ -32130,8 +32249,18 @@ skill_db: (
 		NoDamage: true
 	}
 	SkillInstances: 3
+    SkillData1: 30000
```
Comment (EPuncker) [inline db/re/skill_db.conf:32252]:
```suggestion
	SkillData1: 30000
```

File: db/re/skill_db.conf [hunk 18]
```diff
@@ -32130,8 +32249,18 @@ skill_db: (
...
+    Cooldown: 1000
```
Comment (EPuncker) [inline db/re/skill_db.conf:32253]:
```suggestion
	Cooldown: 1000
```

File: src/map/battle.c [hunk 19]
```diff
@@ -2758,6 +2765,61 @@ static int battle_calc_skillratio(int attack_type, struct block_list *src, struc
...
+					skillratio += -100 + 200 * skill_lv;
+					break;
+				case RL_FIRE_RAIN:
+				case RL_AM_BLAST:
+					skillratio += -100 + 3500 + 300 * skill_lv;
+					break;
+				case RL_HAMMER_OF_GOD:
+					skillratio += -100 + 100 * skill_lv;
+					if (sd) {
+						if (flag & 8)
```
Comment (4144) [inline src/map/battle.c:2797]: should be ``if ((flag & 8) != 0)``

File: src/map/battle.c [hunk 20]
```diff
@@ -2272,6 +2277,8 @@ static int battle_calc_skillratio(int attack_type, struct block_list *src, struc
 					break;
 				case GS_DESPERADO:
 					skillratio += 50 * (skill_lv-1);
+					if (sc && sc->data[SC_FALLEN_ANGEL])
```
Comment (4144) [inline src/map/battle.c:2280]: should be ``sc != NULL && ...``

File: src/map/battle.c [hunk 21]
```diff
@@ -2758,6 +2765,61 @@ static int battle_calc_skillratio(int attack_type, struct block_list *src, struc
 				case MH_MAGMA_FLOW:
 					skillratio += -100 + 100 * skill_lv;
 					break;
+				case RL_FIREDANCE:
+					skillratio += 100 + 100 * skill_lv;
+					skillratio += (sd ? pc->checkskill(sd, GS_DESPERADO) * 20 : 0);
```
Comment (4144) [inline src/map/battle.c:2770]: should be ``sd != NULL`` here and many lines below

File: src/map/battle.c [hunk 22]
```diff
@@ -3280,6 +3342,17 @@ static int64 battle_calc_damage(struct block_list *src, struct block_list *bl, s
 				status_change_end(bl, SC_KYRIE, INVALID_TIMER);
 		}
 
+		if ((sce = sc->data[SC_PLATINUM_ALTER]) && damage > 0) {
```
Comment (4144) [inline src/map/battle.c:3345]: should be ``if ((sce = sc->data[SC_PLATINUM_ALTER]) != NULL && damage > 0) {``

File: src/map/battle.c [hunk 23]
```diff
@@ -4893,6 +4984,17 @@ static struct Damage battle_calc_weapon_attack(struct block_list *src, struct bl
 			case LG_BANISHINGPOINT:
 				hitrate += 3 * skill_lv;
 				break;
+			case RL_SLUGSHOT:
+				{
+					int8 dist = distance_bl(src, target);
```
Comment (4144) [inline src/map/battle.c:4989]: better use ``int`` here

File: src/map/clif.c [hunk 24]
```diff
@@ -24390,6 +24390,24 @@ static bool clif_lapineUpgrade_result(struct map_session_data *sd, enum lapineUp
...
+/**
+* Marks Crimson Marker target on mini-map to the caster
+* 09C1 <id>.L <x>.W <y>.W (ZC_C_MARKERINFO)
+* @param fd
+* @param bl Crimson Marker target
+**/
+void clif_crimson_marker(struct map_session_data *sd, struct block_list *bl, bool remove)
+{
+	unsigned char buf[10];
+	nullpo_retv(sd);
```
Comment (4144) [inline src/map/clif.c:24402]: missing ``nullpo_retv(bl)``

File: src/map/packets.h [hunk 25]
```diff
@@ -1425,6 +1425,7 @@ packet(0x96e,clif->ackmergeitems);
 	packet(0x0887,clif->pStoragePassword,0);  // CZ_ACK_STORE_PASSWORD
 	packet(0x0363,clif->pDull); // CZ_JOIN_BATTLE_FIELD
 	packet(0x0436,clif->pDull); // CZ_GANGSI_RANK
+	packet(0x09C1,clif->crimson_marker,10,2,6,8); // ZC_C_MARKERINFO
```
Comment (4144) [inline src/map/packets.h:1428]: need set correct versions for packet 0x09C1. it exists starting from clients:
2013-07-31cRagexe
2013-07-17cRagexeRE
and any zero

also for packet better use struct in clif_crimson_marker

File: src/map/pc.c [hunk 26]
```diff
@@ -10567,6 +10569,9 @@ static int pc_unequipitem(struct map_session_data *sd, int n, int flag)
 		pc->unequipitem(sd, sd->equip_index[EQI_AMMO], PCUNEQUIPITEM_FORCE);
 #endif
 
+	if (sd->inventory_data[n]->type == IT_AMMO && (sd->inventory_data[n]->nameid != ITEMID_SILVER_BULLET || sd->inventory_data[n]->nameid != ITEMID_SANCTIFIED_BULLET || sd->inventory_data[n]->nameid != ITEMID_SILVER_BULLET_))
```
Comment (4144) [inline src/map/pc.c:10572]: missing check for ``sd->inventory_data[n] != NULL``

File: src/map/skill.c [hunk 27]
```diff
@@ -5910,7 +6004,54 @@ static int skill_castend_damage_id(struct block_list *src, struct block_list *bl
...
+				if (sd && sd->flicker && tsc && tsc->data[SC_HOWLING_MINE] && tsc->data[SC_HOWLING_MINE]->val2 == src->id) {
+					map->foreachinrange(skill->area_sub, bl, skill->get_splash(skill_id, skill_lv), BL_CHAR | BL_SKILL, src, skill_id, skill_lv, tick, flag | BCT_ENEMY | 1, skill->castend_damage_id);
+					flag |= 1; // Don't consume requirement
+					tsc->data[SC_HOWLING_MINE]->val3 = 1; // Mark the SC end because not expired
+					status_change_end(bl, SC_HOWLING_MINE, INVALID_TIMER);
+					sc_start4(src, bl, SC_BURNING, 10 * skill_lv, skill_lv, 1000, src->id, 0, skill->get_time2(skill_id, skill_lv));
+				}
+			}
+			else
+				skill->attack(skill->get_type(skill_id, skill_lv), src, src, bl, skill_id, skill_lv, tick, flag);
```
Comment (4144) [inline src/map/skill.c:6023]: here better put this line into ``{}`` because same was before else

File: src/map/skill.c [hunk 28]
```diff
@@ -5910,7 +6004,54 @@ static int skill_castend_damage_id(struct block_list *src, struct block_list *bl
...
+			if (sd && sd->flicker)
+				flag |= 1; // Don't consume requirement
+			break;
+		case RL_QD_SHOT:
+			if (skill->area_temp[1] == bl->id)
+				break;
+			if (flag & 1 && tsc && tsc->data[SC_CRIMSON_MARKER])
```
Comment (4144) [inline src/map/skill.c:6030]: better put ``(flag & 1) && ...``

File: src/map/skill.c [hunk 29]
```diff
@@ -5910,7 +6004,54 @@ static int skill_castend_damage_id(struct block_list *src, struct block_list *bl
...
+				skill->attack(skill->get_type(skill_id, skill_lv), src, src, bl, skill_id, skill_lv, tick, flag | SD_ANIMATION);
+			break;
+		case RL_D_TAIL:
+		case RL_HAMMER_OF_GOD:
+			if (flag & 1)
+				skill->attack(skill->get_type(skill_id, skill_lv), src, src, bl, skill_id, skill_lv, tick, flag | SD_ANIMATION);
```
Comment (4144) [inline src/map/skill.c:6036]: missing ``{}`` in if block, because ``{}`` used in else block

File: src/map/skill.c [hunk 30]
```diff
@@ -11096,6 +11243,75 @@ static int skill_castend_nodamage_id(struct block_list *src, struct block_list *
 			break;
 		case SO_ELEMENTAL_SHIELD:/* somehow its handled outside this switch, so we need a empty case otherwise default would be triggered. */
 			break;
+		case RL_RICHS_COIN:
+			if (sd) {
+				int i;
+
+				clif->skill_nodamage(src, bl, skill_id, skill_lv, 1);
+				for (i = 0; i < 10; i++)
```
Comment (4144) [inline src/map/skill.c:11251]: should be ``for (int i = 0; i < ...``

File: src/map/skill.c [hunk 31]
```diff
@@ -12274,7 +12491,43 @@ static int skill_castend_pos2(struct block_list *src, int x, int y, uint16 skill
...
+				clif->snap(src, src->x, src->y);
+				sc_start(src, src, type, 100, skill_id, skill->get_time(skill_id, skill_lv));
+			} else {
+				if (sd)
+					clif->skill_fail(sd, skill_id, USESKILL_FAIL_LEVEL, 0, 0);
+			}
+			break;
+		case RL_FIRE_RAIN:
+			{
+				int w, wave = skill_lv + 5, dir = map->calc_dir(src, x, y);
```
Comment (4144) [inline src/map/skill.c:12505]: better assign each variable in own line with int at start

File: src/map/skill.c [hunk 32]
```diff
@@ -12274,7 +12491,43 @@ static int skill_castend_pos2(struct block_list *src, int x, int y, uint16 skill
...
+				int sx = x = src->x, sy = y = src->y;
+
+				for (w = 0; w <= wave; w++) {
```
Comment (4144) [inline src/map/skill.c:12508]: should be ``for (int w = 0; ...``

File: src/map/skill.c [hunk 33]
```diff
@@ -14047,6 +14307,17 @@ static int skill_unit_onplace_timer(struct skill_unit *src, struct block_list *b
...
 			break;
+		case UNT_FIRE_RAIN:
+			clif->skill_damage(ss, bl, tick, status_get_amotion(ss), 0,
+				skill->attack(skill->get_type(sg->skill_id, sg->skill_lv), ss, &src->bl, bl, sg->skill_id, sg->skill_lv, tick, SD_ANIMATION | SD_SPLASH),
+				1, sg->skill_id, sg->skill_lv, BDT_SKILL);
+			break;
+		case UNT_B_TRAP:
+			if (tsc && tsc->data[type])
+				break;
+			sc_start(ss, bl, type, 100, sg->skill_lv, skill_get_time2(sg->skill_id,sg->skill_lv));
```
Comment (4144) [inline src/map/skill.c:14318]: missing space

File: src/map/skill.c [hunk 34]
```diff
@@ -17357,6 +17637,30 @@ static int skill_detonator(struct block_list *bl, va_list ap)
...
+ */
+static int skill_bind_trap(struct block_list *bl, va_list ap) {
+	struct skill_unit *su = NULL;
+	struct block_list *src = NULL;
+
+	nullpo_ret(bl);
+
+	src = va_arg(ap, struct block_list*);
+
+	if (bl->type != BL_SKILL || !(su = (struct skill_unit*)bl) || !(su->group))
```
Comment (4144) [inline src/map/skill.c:17652]: use here ``BL_CAST``

File: src/map/skill.c [hunk 35]
```diff
@@ -17357,6 +17637,30 @@ static int skill_detonator(struct block_list *bl, va_list ap)
...
+ * Rebellion's Bind Trap explosion
+ * @author [Cydh]
+ */
+static int skill_bind_trap(struct block_list *bl, va_list ap) {
+	struct skill_unit *su = NULL;
+	struct block_list *src = NULL;
+
+	nullpo_ret(bl);
+
+	src = va_arg(ap, struct block_list*);
```
Comment (4144) [inline src/map/skill.c:17650]: missing ``nullpo_ret(src)``

File: src/map/skill.c [hunk 36]
```diff
@@ -20293,6 +20623,31 @@ static void skill_init_unit_layout(void)
 		}
 		pos++;
 	}
+	skill->firerain_unit_pos = pos;
+	for (i = 0; i < 8; i++)
+	{
+		skill->dbs->unit_layout[pos].count = 3;
+		switch (i) {
+		case 0: case 1: case 3: case 4: case 5: case 7:
```
Comment (4144) [inline src/map/skill.c:20631]: better put case lables in separate lines

**Review by 4144 — CHANGES_REQUESTED** [review]

File: src/map/pc.c (hunk 26 above)
Comment (MrKeiKun) [reply src/map/pc.c:10572]: oh yeah i never noticed it was outside the `sd->inventory_data[n] != NULL`

File: src/map/skill.c (hunk 30 above)
Comment (MrKeiKun) [reply src/map/skill.c:11251]: lol yeah. i was not into such but since only loop uses it. that should be the case.

File: src/map/pc.c [hunk 37]
```diff
@@ -12608,6 +12615,23 @@ static bool pc_auto_exp_insurance(struct map_session_data *sd)
 	return true;
 }
 
+/**
+* Clear Crimson Marker data from caster
+* @param sd: Player
+**/
+void pc_crimson_marker_clear(struct map_session_data *sd)
+{
+	uint8 i;
```
Comment (4144) [inline src/map/pc.c:12624]: i should be with ``for`` line and type better use int

File: src/map/pc.c [hunk 38]
```diff
@@ -10567,6 +10569,11 @@ static int pc_unequipitem(struct map_session_data *sd, int n, int flag)
 		pc->unequipitem(sd, sd->equip_index[EQI_AMMO], PCUNEQUIPITEM_FORCE);
 #endif
 
+	if (sd->inventory_data[n] != NULL) {
+			if (sd->inventory_data[n]->type == IT_AMMO && (sd->inventory_data[n]->nameid != ITEMID_SILVER_BULLET || sd->inventory_data[n]->nameid != ITEMID_SANCTIFIED_BULLET || sd->inventory_data[n]->nameid != ITEMID_SILVER_BULLET_))
```
Comment (4144) [inline src/map/pc.c:10573]: look like code offset bit wrong

File: src/map/skill.c [hunk 39]
```diff
@@ -5910,7 +6004,54 @@ static int skill_castend_damage_id(struct block_list *src, struct block_list *bl
 		case SU_SCAROFTAROU:
 			sc_start(src, bl, skill->get_sc_type(skill_id), 10, skill_lv, skill->get_time(skill_id, skill_lv)); // TODO: What's the activation chance for the effect?
 			break;
+		case RL_H_MINE:
+			if (!(flag & 1)) {
```
Comment (4144) [inline src/map/skill.c:6008]: on both lines look like need check to 0 and NULL

**Review by 4144 — CHANGES_REQUESTED** [review]

File: src/map/pc.c (hunk 38 above)
Comment (MrKeiKun) [reply src/map/pc.c:10573]: wait what. O_o
VS seems to be bugging my tabbing setup.

File: src/map/skill.c [hunk 40]
```diff
@@ -17357,6 +17634,31 @@ static int skill_detonator(struct block_list *bl, va_list ap)
...
+/**
+ * Rebellion's Bind Trap explosion
+ * @author [Cydh]
+ */
+static int skill_bind_trap(struct block_list *bl, va_list ap) {
+	struct skill_unit *su = NULL;
+	struct block_list *src = NULL;
+
+	nullpo_ret(bl);
+	nullpo_ret(src);
```
Comment (4144) [inline src/map/skill.c:17646]: this line should be after ``src = va_arg(ap, struct block_list*);``

**Review by 4144 — CHANGES_REQUESTED** [review]

**Review by 4144 — APPROVED** [review]

**Review by MishimaHaruna — APPROVED** [review]

**Pull request merged**